Text file type detection updates

jmarshall opened the pull request:
Dusting off some old branches, notably to fix #200, making identification of SAM files more accurate. At present, any text file is recognised as SAM, which is less than ideal.

This adds four new entries to `htsExactFormat`, so it may be a good time to review #587 with a view to adding these entries with a suitable prefix from the start.

Unfortunately some code, notably bcftools, has taken advantage of the bug that any text file is recognised as SAM to open arbitrary text files with `hts_open()` in contravention of its documentation:

```
/*! @abstract       Open a SAM/BAM/CRAM/VCF/BCF/etc file
…
*/
htsFile *hts_open(const char *fn, const char *mode);
```

Ideally such code would be rewritten to use `bgzf_open`/`bgzf_getline`/etc instead of misusing `hts_open`/`hts_getline`/etc.

Alternatively the use of `hts_open` to read text files could be blessed. The patch contains workarounds to enable such use (as currently _tabix.c_ is guilty of the same thing), but bcftools may need similar alterations to that in _tabix.c_ to work with this fixed file type detection PR.

daviesrob commented:
Thanks.  It does cause issues for bcftools, so we'll have to sort those out before merging this.

jmarshall commented:
The bcftools problem is that it has file arguments that are now recognised as `bed` or `fasta`, that it tries to open using `hts_open()` — which fails because those formats are not in the `case` statements in `hts_open()` and `hts_close()`.

So this is a policy question for the maintainers: should `htsFile` continue to be for SAM/BAM/CRAM and VCF/BCF files exclusively?

If so, the fix for bcftools is to do something different for `bed` files in _synced_bcf_reader.c_'s `bcf_sr_regions_init()` (and maybe a couple of places in bcftools), and to apply a fix like samtools/bcftools#821.

If not, the fix for bcftools is to add `fasta`and `bed` and other text formats to those two `case` statements, and bless the use of `htsFile` for use with `hts_getline` and text input files. (And maybe apply that bcftools PR anyway.)

BTW I have a couple more commits to add to this branch while you folks contemplate that policy question…

jkbonfield commented:
What's the proposed alternative to `hts_open`?  I guess `bgzf_open` is the closest thing in functionality.  (Indeed I just realised your other PR uses that.)

I'm not sure what the original intention was for hts_open with regards to policy.  It does claim to be "format-neutral I/O", but doesn't really clarify beyond that.

I've always found it a bit confusing that the htsExactFormat enum has both specific file types (bam, sam, bcf, bed, etc) as well as generic ones (binary, text).  The category also has unknown in there, rather than it being an error.  Both of those do rather imply it is intended for generic data, but it's a bit of guesswork.    If it wasn't intended that way, then I'm not convinced the generic ones should exist at all in something claiming to be "exact" format,  although internally they do sometimes get used.  (Eg VCF is sometimes labelled as text and not vcf; I fixed this in the indexing PR.)

So my gut instinct is it should probably work.  Breaking other code though implies this is perhaps best delayed until the ABI breaking stuff lands (even though this is only API breaking, it's easy to lump them together and the timing isn't great for a late change like this).

jmarshall commented:
The proposed alternative is

> Ideally such code would be rewritten to use `bgzf_open`/`bgzf_getline`/etc instead

…or plain `hopen`/`hgetln`/etc if they don't need to take advantage of gzip/BGZF decompression being done for them.

`text_format`/`binary_format` exist because after `hts_open("wb")` it's impossible to know whether the exact format is going to be `bam` or `bcf` — see 529ca88071e775a488e4108945d5bd863a4190f0. This is called compromising with reality :smile: But in general `hts_detect_format()` is intended to work on any file, which is not necessarily the case for `hts_open()`. The two are separate.

It turns out that that “_hts_open: Open a SAM/BAM/CRAM/VCF/BCF/etc file_” documentation was written by me in 2013 in d038d8ed33a850b1ba2e634d43cd61fb135a1800. The `hts_getline` function was already present at the time (though it was introduced for reading SAM files). So one point of view is that the 2013 documentation over-constrained the function and it would be in keeping with the function as it existed prior to 2013 to relax that to “_Open a sequence data (SAM/BAM/CRAM) or variant data (VCF/BCF) or textual line-orientated file_” (or something shorter!) and thus bless reading line-by-line with `hts_open()`.

If people agree with this solution, I can rework this PR to avoid any breakage, in bcftools or otherwise.

> … until the ABI breaking stuff lands … timing … a late change …

Would you like to give us a hint when ABI breakage is planned to land, and what timing “late” might relate to?

jkbonfield commented:
I see, I hadn't realised the distinction between what hts_detect_format is meant to work on vs hts_open.  None of this is particularly obvious!

As far as timing, the release candidate is meant to be today, so this is what I mean by late. :-)
The ABI breaking stuff got punted again sadly, but the plan is it'll arrive asap after the release goes out, so develop has the new stuff in for the longest time it can, to give us lots of time to spot potential gotchas.

jmarshall commented:
I'm planning to rework this as described in https://github.com/samtools/htslib/pull/721#issuecomment-400666150 so as to minimise other changes needed.

jmarshall commented:
This branch is now finally updated to a point where it is ready for merging. Rebased and conflicts resolved, and two significant changes from previously which mean it is now ready:

* Disallowing empty SAM files (#261) is now implemented: empty files can be opened with `hts_open()` and read with `hts_getline`, but `sam_hdr_read` now fails. Why it was necessary to implement this is now lost in the mists of time, but it has to do with the _xx#blank.sam_ test cases involving empty files — without work one way or the other in this area, the test suite was failing with this PR.

* Reworked to bless using `hts_open` to read arbitrary text files line-by-line, and tests added to ensure that this works. With this done, there are no issues caused for bcftools.

All of the htslib, samtools, and bcftools test suites pass when run against this branch.

jkbonfield commented:
The xx#blank.sam test was one of mine, lifted over from io_lib.  It was deliberately empty in order to test that case, as it's a valid file, but I'm completely in agreement with taking a stance that valid-or-not, we shouldn't accept it.  The fudge to put a comment in it is good as it still means it tests functionality in e.g. iterators.

Thanks.

valeriuo commented on test/sam.c:
```diff
@@ -1043,6 +1043,41 @@ static void faidx1(const char *filename)
+{
+    samFile *in = sam_open(filename, "r");
+    if (in) {
+        bam1_t *aln = bam_init1();
```
A check for the _empty_format_ would be a nice addition.
Something like:
```c
if (in->format.format != empty_format)
    fail("file %s is not marked as empty\n", filename);
```

jmarshall replied:
Good idea, thanks.

valeriuo commented:
Merged into develop
